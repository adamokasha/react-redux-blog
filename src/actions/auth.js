import axios from "axios";

export const login = (displayName, id, role) => ({
  type: "LOGIN",
  displayName,
  id,
  role
});

export const startSignup = (displayName, email, password) => {
  return async dispatch => {
    try {
      const response = await axios.post(`${process.env.API_URI}/users`, {
        displayName,
        email,
        password
      });

      const user = {
        displayName: response.data.displayName,
        id: response.data._id,
        token: response.headers["x-auth"],
        role: response.data.role
      };
      localStorage.setItem("auth", JSON.stringify(user));

      dispatch(
        login(response.data.displayName, response.data._id, response.data.role)
      );
    } catch (e) {
      throw e.response.data.errors;
    }
  };
};

export const startLogin = (email, password) => {
  return async dispatch => {
    try {
      const response = await axios.post(`${process.env.API_URI}/users/login`, {
        email,
        password
      });

      const user = {
        displayName: response.data.displayName,
        id: response.data._id,
        token: response.headers["x-auth"],
        role: response.data.role
      };
      localStorage.setItem("auth", JSON.stringify(user));

      dispatch(
        login(response.data.displayName, response.data._id, response.data.role)
      );
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const startLogout = () => {
  return async dispatch => {
    try {
      const { token } = JSON.parse(localStorage.getItem("auth"));
      localStorage.removeItem("auth");
      await axios.delete(`${process.env.API_URI}/users/me/token`, {
        headers: {
          "x-auth": token
        }
      });
      dispatch(logout());
    } catch (e) {
      localStorage.removeItem("auth");
      dispatch(logout());
    }
  };
};

export const logout = () => ({
  type: "LOGOUT"
});
