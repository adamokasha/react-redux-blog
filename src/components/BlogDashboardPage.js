import React from "react";
import { connect } from "react-redux";

import Header from "./Header";
import PostsList from "./PostsList";

export const BlogDashboardPage = ({ displayName }) => {
  function renderWelcome() {
    if (displayName) {
      return <h3>Welcome {displayName}.</h3>;
    }
    return;
  }
  return (
    <div>
      <Header />
      <div id="landing">
        <div className="primary-overlay">
          <div className="container h-100">
            <div className="row hero-text">
              <div className="col-lg-7">
                <h1 className="text-white mb-5">
                  React Blog - Responsive blog styled with Bootstrap 4
                </h1>
                <a
                  href="https://github.com/adamokasha/react-redux-blog"
                  target="_blank"
                  className="text-uppercase text-white hero-link"
                >
                  <u>Learn More</u>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <br />
      <div className="container mb-3">{renderWelcome()}</div>
      <div>
        <PostsList />
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  displayName: state.auth.displayName
});

export default connect(mapStateToProps)(BlogDashboardPage);
