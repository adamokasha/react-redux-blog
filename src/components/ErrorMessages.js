import React from 'react';

class ErrorMessages extends React.Component {
  constructor(props) {
    super(props);
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  render() {
    return (
      <ul className="error-list list-group mb-3">
        {this.props.errors.map(error => {
          return (
            <li className="list-group-item list-group-item-danger">{error}</li>
          );
        })}
      </ul>
    );
  }
}

export default ErrorMessages;
