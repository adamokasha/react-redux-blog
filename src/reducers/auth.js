const defaultAuthState = () => {
  if (localStorage.getItem("auth")) {
    return JSON.parse(localStorage.getItem("auth"));
  } else {
    return {};
  }
};

const authReducer = (state = defaultAuthState(), action) => {
  switch (action.type) {
    case "LOGIN":
      return {
        displayName: action.displayName,
        id: action.id,
        role: action.role
      };
    case "LOGOUT":
      return {};
    default:
      return state;
  }
};

export default authReducer;
