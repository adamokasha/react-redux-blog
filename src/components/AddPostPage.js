import React from "react";
import { connect } from "react-redux";

import PostForm from "./PostForm";
import { startAddPost } from "../actions/posts";

export class AddPostPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: ""
    };
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  renderError = () => {
    if (this.state.error) {
      return (
        <div className="error-msg alert alert-danger" role="alert">
          <p>{this.state.error}</p>
        </div>
      );
    }
  };
  onSubmit = async post => {
    try {
      await this.props.startAddPost(post);
      this.props.history.push("/");
    } catch (e) {
      this.setState({ error: e.message });
    }
  };
  render() {
    return (
      <div className="container">
        <h3 className="text-center mt-4">Add Post</h3>
        {this.renderError()}
        <PostForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post))
});

export default connect(
  null,
  mapDispatchToProps
)(AddPostPage);
