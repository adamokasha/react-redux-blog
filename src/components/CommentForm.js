import React from "react";
import { connect } from "react-redux";

import { startAddComment } from "../actions/posts";

export class CommentForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      comment: "",
      error: ""
    };
  }
  onCommentChange = e => {
    this.setState({ comment: e.target.value });
  };
  renderError = () => {
    if (this.state.error) {
      return <div className="error-msg text-danger">{this.state.error}</div>;
    }
  };
  onSubmit = e => {
    e.preventDefault();

    const { comment } = this.state;
    if (comment.length < 8 || comment.length > 128) {
      this.setState({ error: "Comment must be between 8 and 128 characters." });
    } else {
      this.setState({ error: "" });
      this.props.startAddComment(this.props.id, this.state.comment);
      this.setState({ comment: "" });
    }
  };
  render() {
    return (
      <div className="mb-3">
        {this.renderError()}
        <form onSubmit={this.onSubmit}>
          <div className="form-group col-md-4 pl-0 mb-2">
            <textarea
              className="form-control"
              id="comment-textarea"
              placeholder="Leave a comment..."
              onChange={this.onCommentChange}
              maxLength={128}
              rows="3"
              value={this.state.comment}
            />
          </div>
          <button className="btn btn-primary">Add Comment</button>
        </form>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddComment: (id, comment) => dispatch(startAddComment(id, comment))
});

export default connect(
  null,
  mapDispatchToProps
)(CommentForm);
