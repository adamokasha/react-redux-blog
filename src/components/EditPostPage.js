import React from "react";
import { connect } from "react-redux";

import PostForm from "./PostForm";
import { startEditPost } from "../actions/posts";
import { startDeletePost } from "../actions/posts";

export class EditPostPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: ""
    };
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  renderError = () => {
    if (this.state.error) {
      return (
        <div className="error-msg">
          <p>{this.state.error}</p>
        </div>
      );
    }
  };
  onSubmit = async post => {
    try {
      await this.props.startEditPost(post, this.props.post._id);
      this.props.history.push("/");
    } catch (e) {
      this.setState({ error: e.message });
    }
  };
  deletePost = async () => {
    try {
      await this.props.startDeletePost(this.props.post._id);
      this.props.history.push("/");
    } catch (e) {
      this.setState({ error: e.message });
    }
  };
  render() {
    return (
      <div className="container">
        <h3 className="text-center mt-4">Edit Post</h3>
        <button
          className="btn-delete btn btn-dark btn-sm float-right"
          onClick={this.deletePost}
        >
          <i className="far fa-trash-alt mr-2" />
          Delete Post
        </button>
        {this.renderError()}
        <PostForm post={this.props.post} onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  post: state.posts.find(post => post._id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  startEditPost: (post, id) => dispatch(startEditPost(post, id)),
  startDeletePost: id => dispatch(startDeletePost(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditPostPage);
