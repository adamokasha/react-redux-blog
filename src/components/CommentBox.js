import React from "react";
import { connect } from "react-redux";
import moment from "moment";

import CommentForm from "./CommentForm";
import { sortCommentsByDate } from "../selectors/posts";

export class CommentBox extends React.Component {
  renderCommentForm = () => {
    if (this.props.auth.id) {
      return <CommentForm id={this.props.id} />;
    }
    return <p className="font-italic">You must be signed in to comment...</p>;
  };
  renderComments = () => {
    if (this.props.comments.length > 0) {
      return sortCommentsByDate(this.props.comments).map(comment => {
        return (
          <li className="list-group-item pl-0" key={comment._id}>
            <p className="text-muted mb-1">
              Posted by {comment.createdBy} on{" "}
              {moment(comment.date).format("MM-DD-YY [at] HH:mm")}
            </p>
            <p className="mb-1">{comment.comment}</p>
          </li>
        );
      });
    } else {
      return <p>Be the first to comment...</p>;
    }
  };
  render() {
    return (
      <div>
        <h4>Comments</h4>
        {this.renderCommentForm()}
        <ul className="list-group-flush pl-0">{this.renderComments()}</ul>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(mapStateToProps)(CommentBox);
