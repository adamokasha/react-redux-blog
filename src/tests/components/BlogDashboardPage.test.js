import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { BlogDashboardPage } from "../../components/BlogDashboardPage";
import { withUserRole } from "../fixtures/auth";

configure({ adapter: new Adapter() });

test("should render ExpenseDashBoardPage correctly", () => {
  const wrapper = shallow(
    <BlogDashboardPage displayName={withUserRole.displayName} />
  );
  expect(wrapper).toMatchSnapshot();
});
