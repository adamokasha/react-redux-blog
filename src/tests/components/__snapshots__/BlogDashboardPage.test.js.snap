// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseDashBoardPage correctly 1`] = `
<div>
  <Connect(Header) />
  <div
    id="landing"
  >
    <div
      className="primary-overlay"
    >
      <div
        className="container h-100"
      >
        <div
          className="row hero-text"
        >
          <div
            className="col-lg-7"
          >
            <h1
              className="text-white mb-5"
            >
              React Blog - Responsive blog styled with Bootstrap 4
            </h1>
            <a
              className="text-uppercase text-white hero-link"
              href="https://github.com/adamokasha/react-redux-blog"
              target="_blank"
            >
              <u>
                Learn More
              </u>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
  <br />
  <div
    className="container mb-3"
  >
    <h3>
      Welcome 
      theuser
      .
    </h3>
  </div>
  <div>
    <Connect(PostsList) />
  </div>
</div>
`;
